File: opijae/leetcode/20_Valid_Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        for c in s:
            # 열린 괄호면 스택에 넣기
            if c =='(' or c =='[' or c =='{':
                stack.append(c)
                continue
            if len(stack)==0:
                return False
            
            temp = stack.pop()
            # 닫힌 괄호면 pop한 값이랑 비교
            if temp=='(':
                if c !=')':
                    return False
            if temp=='[':
                if c !=']':
                    return False
            if temp=='{':
                if c !='}':
                    return False
        if len(stack)>0:
            return False
        return True

File: sally/leetcode/20_ Valid Parentheses.cpp
// 20. Valid Parentheses
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Valid Parentheses.
// Memory Usage: 6.5 MB, less than 10.15% of C++ online submissions for Valid Parentheses.

#include <stack>
#include <string>

class Solution {
public:
    bool isValid(string s) {
        char opened_b[3] = {'[','{','('};
        char closed_b[3] = {']','}',')'};
        stack<int> st;
        int s_len = s.size();
        
        for(int i = 0; i < s_len; i++){
            bool opened = false;
            for(int j = 0; j < 3; j++)
                if(s[i] == opened_b[j]){
                    opened = true;
                    st.push(j);
                    break;
                }
            if(opened) continue;
            for(int j = 0; j < 3; j++)
                if(s[i] == closed_b[j]){
                    if(st.empty()) return false;
                    if(st.top() == j) st.pop();
                    else return false;
                }
        }
        if (st.empty()) return true;
        else return false;
    }
};